recipe/use_new_cudatoolkit.patch

```diff
@@ -0,0 +1,350 @@
+                cuda/cuda_dlib.cu 
+                cuda/cudnn_dlibapi.cpp
+@@ -786,16 +628,16 @@ if (NOT TARGET dlib)
+                cuda/cuda_data_ptr.cpp
+                cuda/gpu_data.cpp
+                )
+-            list (APPEND dlib_needed_private_libraries ${CUDA_CUBLAS_LIBRARIES})
+-            list (APPEND dlib_needed_private_libraries ${cudnn})
+-            list (APPEND dlib_needed_private_libraries ${CUDA_curand_LIBRARY})
+-            list (APPEND dlib_needed_private_libraries ${CUDA_cusolver_LIBRARY})
+-            list (APPEND dlib_needed_private_libraries ${CUDA_CUDART_LIBRARY})
++            list (APPEND dlib_needed_private_libraries CUDA::cublas)
++            list (APPEND dlib_needed_private_libraries ${CUDNN_LIBRARY_PATH})
++            list (APPEND dlib_needed_private_libraries CUDA::curand)
++            list (APPEND dlib_needed_private_libraries CUDA::cusolver)
++            list (APPEND dlib_needed_private_libraries CUDA::cudart)
+             if(openmp_libraries)
+                list (APPEND dlib_needed_private_libraries ${openmp_libraries})
+             endif()
+ 
+-            include_directories(${cudnn_include})
++            include_directories(${CUDAToolkit_INCLUDE_DIRS})
+             message(STATUS "Enabling CUDA support for dlib.  DLIB WILL USE CUDA, compute capabilities: ${DLIB_CUDA_COMPUTE_CAPABILITIES}")
+          else()
+             set(DLIB_USE_CUDA OFF CACHE STRING ${DLIB_USE_BLAS_STR} FORCE )
+@@ -851,15 +693,7 @@ if (NOT TARGET dlib)
+          endif()
+       endif()
+ 
+-      # Tell CMake to build dlib via add_library()/cuda_add_library()
+-      if (DLIB_USE_CUDA)
+-         # The old cuda_add_library() command doesn't support CMake's newer dependency
+-         # stuff, so we have to set the include path manually still, which we do here.
+-         include_directories(${dlib_needed_public_includes})
+-         cuda_add_library(dlib ${source_files} )
+-      else()
+-         add_library(dlib ${source_files} )
+-      endif()
++      add_library(dlib ${source_files})
+ 
+    endif ()  ##### end of if NOT DLIB_ISO_CPP_ONLY ##########################################################
+ 
+diff --git a/dlib/cmake_utils/FindCUDNN.cmake b/dlib/cmake_utils/FindCUDNN.cmake
+new file mode 100644
+index 00000000..82134328
+--- /dev/null
++++ b/dlib/cmake_utils/FindCUDNN.cmake
+@@ -0,0 +1,78 @@
++# Find the CUDNN libraries
++#
++# The following variables are optionally searched for defaults
++#  CUDNN_ROOT: Base directory where CUDNN is found
++#  CUDNN_INCLUDE_DIR: Directory where CUDNN header is searched for
++#  CUDNN_LIBRARY: Directory where CUDNN library is searched for
++#  CUDNN_STATIC: Are we looking for a static library? (default: no)
++#
++# The following are set after configuration is done:
++#  CUDNN_FOUND
++#  CUDNN_INCLUDE_PATH
++#  CUDNN_LIBRARY_PATH
++#
++
++include(FindPackageHandleStandardArgs)
++
++set(CUDNN_ROOT $ENV{CUDNN_ROOT_DIR} CACHE PATH "Folder containing NVIDIA cuDNN")
```

Comment:
Am seeing the following error [on CI]( https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=854335&view=logs&j=ee40e33f-3c6c-5798-f3bb-a5cc2626bba8&t=a449e4ba-8c34-57b8-d8bf-0a22a3446d56 ), which suggest build can't find cuDNN:

```
  D:\bld\dlib-split_1704676638831\work\dlib\cuda\cudnn_dlibapi.cpp(10,10): fatal error C1083: Cannot open include file: 'cudnn.h': No such file or directory [D:\bld\dlib-split_1704676638831\work\build\temp.win-amd64-cpython-39\Release\dlib_build\dlib.vcxproj]
```

It looks like this patch lets us set where to look for cuDNN with `CUDNN_ROOT_DIR`. Have we tried modifying `bld.bat` to set `CUDNN_ROOT_DIR` to `LIBRARY_PREFIX`, which is [where the package puts these]( https://github.com/conda-forge/cudnn-feedstock/blob/d6ac758743c251afd1be583fa4c22a53baac4a17/recipe/meta.yaml#L37-L42 )?

Reply:
Ah nice, I didn't even notice that some of the builds are getting somewhere. The CUDA12 build is still reporting that the CUDAToolkit could not be found. The CUDA11.2 and CUDA11.8 build errors differ as well. I've added your (slightly modified) suggestion, let's see what happens.

Reply:
Ok we may need to [set `CUDAToolkit_ROOT`]( https://cmake.org/cmake/help/latest/module/FindCUDAToolkit.html#search-behavior ) as well

Reply:
Should this be set to `CUDA_HOME`?

Reply:
In CUDA 12.0, it is just `LIBRARY_PREFIX` so `CUDA_HOME` is undefined

Setting `CUDA_HOME` in the past was done to workaround CUDA Toolkit not being in the environment